sw/host/opentitanlib/src/io/gpio.rs

```diff
@@ -200,3 +209,36 @@ pub trait GpioMonitoring {
         continue_monitoring: bool,
     ) -> Result<MonitoringReadResponse>;
 }
+
+/// Represents one entry in the specification of a bitbanging waveform.
+pub enum BitbangEntry<'rd, 'wr> {
```

Comment:
There are some slight ambiguities that would be good to clarify: it seems that `Both` does not touch pins in `Input` mode when writing and ignores pins in `Output` mode when reading? But what about `Write`:
does that actively change the gpio pin mode to `Output` for every pin? What happens if a pin is in `Input` mode?

Reply:
`Both` is meant to work the same way as FTDI "synchronous bitbanging", meaning that at each clock tick, one byte from the `'wr` data is applied to the set of pins, (but has no effect on `Input` pins), and also, just before applying, each of the pins (both `Input`, `OpenDrain` and `PushPull`) are sampled, and the value sent back as a byte in the `'rd` data.  (This returned data is not useful for `PushPull` pins, as it would always equal what was sent the tick before, barring a drive fight.)
`Write` does exactly the same thing, except it discards the returned data.
I have elaborated in the comments, please let me know if the explanation is still ambiguous.  (Sometimes when you have been thinking about something for weeks, some aspects may seem obvious to you, while they are not to others.)